### PR TITLE
Add resetter

### PR DESCRIPTION
**Pull request by tvkhoa**
@tungv Add resetter into makefetchAction

### CODE REVIEW

File: src/__tests__/reducer.spec.js [hunk 1]
```diff
@@ -210,4 +210,93 @@ describe('reducer', () => {
...
+          lastResponse: undefined,
+          error: undefined,
+        }
+      };
+      console.log(actual);
+      expect(actual).to.deep.equal(expected);
+    });
+
+    it('should update local with array of all fields', () => {
+      const actual = setup(
```
Comment (tungv) [inline src/__tests__/reducer.spec.js:277]: by "reset all" you probably want to remove `SAMPLE` from state altogether

File: src/makeFetchAction.js [hunk 2]
```diff
@@ -1,5 +1,28 @@
 import get from './utils/get';
-import { CALL_API, REDUCER_PATH, ACTION_UPDATE_LOCAL } from './constants';
+import {
+  CALL_API,
+  REDUCER_PATH,
+  ACTION_UPDATE_LOCAL,
+  ACTION_RESET_LOCAL,
+} from './constants';
+
+const normalizeResetData = data => {
```
Comment (tungv) [inline src/makeFetchAction.js]: Let's use the default parameter syntax instead and move defaultData to the outer scope. Maybe we should set it as default params for `resetter` itself (line 43)

File: src/makeFetchAction.js [hunk 3]
```diff
@@ -1,5 +1,28 @@
...
+  const defaultData = [
+    'lastRequest',
+    'isFetching',
+    'isInvalidated',
+    'lastResponse',
+    'data',
+    'error',
+  ];
+  if (data === undefined) {
```
Comment (tungv) [inline src/makeFetchAction.js]: what if they pass null?

File: src/reducer.js [hunk 4]
```diff
@@ -51,6 +58,22 @@ const reducer = handleActions({
...
       data: action.payload.data,
     }),
+  [ACTION_RESET_LOCAL]: (state, action) => {
+    const name = getName(action);
+    const currentData = state[name] || {};
+    return updateWith(
+      state,
+      name,
+      {
+        lastRequest: resetOrKeepValue('lastRequest', action, currentData),
```
Comment (tungv) [inline src/reducer.js:68]: maybe we can use `lodash/omit` for sake of readability

**Review by tungv — COMMENTED** [review]
I think when they call `resetter()` without params, we can just send an empty `data` or a symbol or something special to let the reducer to remove the API root altogether

File: src/makeFetchAction.js (hunk 3 above)
Comment (tvkhoa) [reply src/makeFetchAction.js]: nothing happened

File: src/__tests__/reducer.spec.js (hunk 1 above)
Comment (tvkhoa) [reply src/__tests__/reducer.spec.js:277]: good idea.

File: src/reducer.js [hunk 5]
```diff
@@ -13,6 +14,12 @@ const getJSONResponse = action => action.payload.json;
 const getError = action => action.payload.error;
 const getPreviousError = (state, action) => state[getName(action)] ? state[getName(action)].error : null;
 
+const includeString = (element, array) => array.indexOf(element) !== -1;
+const resetOrKeepValue = (field, action, currentData) => (
+  includeString(field, action.payload.data) ? undefined : currentData[field]
```
Comment (tungv) [inline src/reducer.js:19]: if they call `resetter(null)` this will break, won't it?

Comment (tungv) [reply src/reducer.js:19]: similar with `false` or `0`. I think in normalizeData function, let check `!Array.isArray(data)` instead

File: src/__tests__/reducer.spec.js (hunk 1 above)
Comment (tvkhoa) [reply src/__tests__/reducer.spec.js:277]: Actually, I think we can set all fields as undefined. It's acceptable

File: src/makeFetchAction.js (hunk 2 above)
Comment (tvkhoa) [reply src/makeFetchAction.js]: Done

File: src/reducer.js (hunk 4 above)
Comment (tvkhoa) [reply src/reducer.js:68]: I think we dont want to use lodash

File: src/reducer.js (hunk 5 above)
Comment (tvkhoa) [reply src/reducer.js:19]: Done

**Pull request merged**